packages/base/query-field-support.ts

```diff
@@ -0,0 +1,304 @@
+      return;
+    }
+    cache = new Map();
+    queryFieldStates.set(instance, cache);
+  }
+
+  if (state) {
+    cache.set(fieldName, state);
+  } else {
+    cache.delete(fieldName);
+    if (cache.size === 0) {
+      queryFieldStates.delete(instance);
+    }
+  }
+}
+
+export function getQueryFieldState(
+  instance: BaseDef,
+  fieldName: string,
+): QueryFieldState | undefined {
+  return queryFieldStates.get(instance)?.get(fieldName);
+}
+
+function getQueryFieldStateKeys(instance: BaseDef): string[] {
+  return Array.from(queryFieldStates.get(instance)?.keys() ?? []);
+}
+
+export function beginQueryFieldEvaluation(
+  instance: BaseDef,
+  fieldName: string,
+): () => void {
+  let guards = queryFieldEvaluationGuards.get(instance);
+  if (!guards) {
+    guards = new Map();
+    queryFieldEvaluationGuards.set(instance, guards);
+  }
+  let current = guards.get(fieldName) ?? 0;
+  guards.set(fieldName, current + 1);
+  return () => {
+    let existing = guards?.get(fieldName);
+    if (existing === undefined) {
+      return;
+    }
+    if (existing <= 1) {
+      guards?.delete(fieldName);
+      if (guards && guards.size === 0) {
+        queryFieldEvaluationGuards.delete(instance);
+      }
+    } else {
+      guards?.set(fieldName, existing - 1);
+    }
+  };
+}
+
+export function markQueryFieldStaleInternal(
+  instance: BaseDef,
+  fieldName: string,
+  notifyCardTracking: (instance: BaseDef) => void,
+): boolean {
+  let state = getQueryFieldState(instance, fieldName);
+  if (!state || state.stale) {
+    return false;
+  }
+  setQueryFieldState(instance, fieldName, {
+    ...state,
+    stale: true,
+  });
+  notifyCardTracking(instance);
+  return true;
+}
+
+export function isQueryFieldEvaluationInProgress(
+  instance: BaseDef,
+  fieldName: string,
+): boolean {
+  return (queryFieldEvaluationGuards.get(instance)?.get(fieldName) ?? 0) > 0;
+}
+
+export function validateRelationshipQuery(
+  ownerPrototype: BaseDef | undefined,
+  fieldName: string,
+  query: Query,
+): void {
+  if (typeof query !== 'object' || query == null) {
+    throw new Error(`query field "${fieldName}" must provide a query object`);
+  }
+  if (!ownerPrototype) {
+    return;
+  }
+  let tokens = new Set<string>();
+  collectInterpolationTokens(query, tokens);
+  let ownerClass = ownerPrototype.constructor as typeof BaseDef;
+  for (let token of tokens) {
+    if (token === THIS_REALM_TOKEN) {
+      continue;
+    }
+    if (
+      typeof token === 'string' &&
+      token.startsWith(THIS_INTERPOLATION_PREFIX)
+    ) {
+      let path = token.slice(THIS_INTERPOLATION_PREFIX.length);
+      let [head] = path.split('.');
+      let referencedField = getField(ownerClass, head, { untracked: true });
+      if (!referencedField) {
+        let ownerName = ownerClass.name ?? 'Card';
+        throw new Error(
+          `query field "${fieldName}" references unknown path "${token}" on ${ownerName}`,
+        );
+      }
+    }
+  }
+}
```

Comment:
Let's make this fully validate by walking down into additional "." segments using getField.